lib/directory.coffee

```diff
@@ -163,34 +165,63 @@ class Directory
       names = []
     names.sort(new Intl.Collator(undefined, {numeric: true, sensitivity: "base"}).compare)
 
-    files = []
-    directories = []
-
+    namePromises = []
     for name in names
-      fullPath = path.join(@path, name)
-      continue if @isPathIgnored(fullPath)
-
-      stat = fs.lstatSyncNoException(fullPath)
-      symlink = stat.isSymbolicLink?()
-      stat = fs.statSyncNoException(fullPath) if symlink
-
-      if stat.isDirectory?()
-        if @entries.hasOwnProperty(name)
-          # push a placeholder since this entry already exists but this helps
-          # track the insertion index for the created views
-          directories.push(name)
-        else
-          expansionState = @expansionState.entries[name]
-          directories.push(new Directory({name, fullPath, symlink, expansionState, @ignoredPatterns}))
-      else if stat.isFile?()
-        if @entries.hasOwnProperty(name)
-          # push a placeholder since this entry already exists but this helps
-          # track the insertion index for the created views
-          files.push(name)
-        else
-          files.push(new File({name, fullPath, symlink, realpathCache}))
-
-    @sortEntries(directories.concat(files))
+      # throw this in a closure so we don't have scoping issues
+      # async/await would make this more pleasant when this is
+      # inevitably ported to ES6
+      f = =>
```

Comment:
Would Coffee's `do` syntax help here at all for an anonymous immediately-invoked function?